File: java17/src/main/java/com/simplify/java17/book/Book.java
package com.simplify.java17.book;

public sealed interface Book permits Book.FictionBook, Book.ProgrammingBook, Book.PsychologicalBook {

  String title();

  String author();

  record FictionBook(String title, String author) implements Book {
    public boolean available() {
      return false;
    }
  }

  record ProgrammingBook(String title, String author) implements Book {
    public boolean exists() {
      return true;
    }
  }

  record PsychologicalBook(String title, String author) implements Book {
    public boolean imAAvailable() {
      return false;
    }
  }

}

File: ktlin/src/main/kotlin/com/simplify/ktlin/book/Book.kt
package com.simplify.ktlin.book

sealed class Book(val title: String, val author: String) {

    class FictionBook(title: String, author: String) : Book(title, author) {
        fun available(): Boolean {
            return false
        }
    }

    class ProgrammingBook(title: String, author: String) : Book(title, author) {
        fun exists(): Boolean {
            return true
        }
    }

    class PsychologicalBook(title: String, author: String) : Book(title, author) {
        fun imAAvailable(): Boolean {
            return false
        }
    }
}